fbpic/lpa_utils/laser/laser.py

```diff
@@ -85,84 +145,11 @@ def add_laser( sim, a0, w0, ctau, z0, zf=None, lambda0=0.8e-6,
-
-    # Handle the introduction method of the laser
-    if method == 'direct':
-        # Directly add the laser to the interpolation object
-        add_laser_direct( sim.fld, E0, w0, ctau, z0, zf, k0, cep_phase,
-            phi2_chirp, theta_pol, fw_propagating, update_spectral, boost )
-    elif method == 'antenna':
-        dr = sim.fld.interp[0].dr
-        Nr = sim.fld.interp[0].Nr
-        Nm = sim.fld.Nm
-        # Add a laser antenna to the simulation object
-        sim.laser_antennas.append(
-            LaserAntenna( E0, w0, ctau, z0, zf, k0, cep_phase,
-                phi2_chirp, theta_pol, z0_antenna, dr, Nr, Nm, boost=boost ) )
-    else:
-        raise ValueError('Unknown laser method: %s' %method)
-
-
-def add_laser_direct( fld, E0, w0, ctau, z0, zf, k0, cep_phase,
-    phi2_chirp, theta_pol, fw_propagating, update_spectral, boost ):
-    """
-    Add a linearly-polarized, Gaussian laser pulse to the Fields object
-
-    Parameters:
-    -----------
-    See the parameters of add_laser
-    NB: all laser quantities are still given in the lab-frame at this point.
-    """
-    # Get the polarization component
-    # (Due to the Fourier transform along theta, the
-    # polarization angle becomes a complex phase in the fields)
-    exptheta = np.exp(1.j*theta_pol)
-    # Sign for the propagation (1 for forward propagation, and -1 otherwise)
-    prop = 2*int(fw_propagating) - 1.
-
-    # When running in the boosted frame, convert the value of the electric
-    # field (the other laser quantities are converted in 'gaussian_profile')
-    if boost is not None:
-        E0, = boost.wavenumber([ E0 ])
-
-    # Get the 2D mesh for z and r
-    # (When running a simulation in boosted frame, then z is the coordinate
-    # in the boosted frame -- if the Fields object was correctly initialized.)
-    r, z = np.meshgrid( fld.interp[1].r, fld.interp[1].z )
-    # Calculate the laser profile on the mesh
-    profile_Eperp, profile_Ez = gaussian_profile( z, r, 0,
-            w0, ctau, z0, zf, k0, cep_phase, phi2_chirp, prop=prop,
-            boost=boost, output_Ez_profile=True )
-
-    # Add the Er and Et fields to the mode m=1 (linearly polarized laser)
-    # (The factor 0.5 is related to the fact that there is a factor 2
-    # in the gathering function, for the particles)
-    fld.interp[1].Er +=  0.5  * E0 * exptheta * profile_Eperp
-    fld.interp[1].Et += -0.5j * E0 * exptheta * profile_Eperp
-    fld.interp[1].Br +=  0.5j * prop * E0/c * exptheta * profile_Eperp
-    fld.interp[1].Bt +=  0.5  * prop * E0/c * exptheta * profile_Eperp
-
-    # Add the Ez fields to the mode m=1 (linearly polarized laser)
-    fld.interp[1].Ez +=  0.5  * E0 * exptheta * profile_Ez
-    fld.interp[1].Bz +=  0.5j * prop * E0/c * exptheta * profile_Ez
-
-    # Up to now, only the interpolation grid was modified.
-    # Now convert the fields to spectral space.
-    if update_spectral:
-        fld.interp2spect('E')
-        fld.interp2spect('B')
```

Comment:
The function `add_laser` is kept for backward-compatibility, but is now much shorter. 

Also, it illustrates how one would initialize a laser with the new API. This API could still change before the next release.

Reply:
Should we maybe already change the example scripts to this API or do you want to wait until we decided on a final API?

Reply:
Yeah, I would rather wait until we have a definitive API. This PR just introduces the necessary mechanism for separating the laser profile from the injection, but not much thought was given to the API.